File: task1/.java
import java.util.*;

class Solution {
    public String minWindow(String s, String t) {
        if (t.isEmpty()) return "";

        Map<Character, Integer> need = new HashMap<>();
        Map<Character, Integer> window = new HashMap<>();

        // Заполняем карту need: ключ - символ из t, значение - частота
        for (char c : t.toCharArray()) {
            need.put(c, need.getOrDefault(c, 0) + 1);
        }

        int needCount = need.size();  // Количество уникальных символов в t
        int formed = 0;  // Количество символов, для которых выполнено need.get(c) == window.get(c)
        int left = 0, right = 0; // Указатели на начало и конец "окна"
        int minLen = Integer.MAX_VALUE; // Минимальная длина найденной подстроки
        String minStr = ""; // Наименьшая подстрока

        while (right < s.length()) {
            char c = s.charAt(right);
            window.put(c, window.getOrDefault(c, 0) + 1);

            // Если символ c из s есть в t и его частота в окне равна требуемой
            if (need.containsKey(c) && window.get(c).equals(need.get(c))) {
                formed++;
            }

            // Пока все необходимые символы найдены
            while (formed == needCount) {
                // Обновляем минимальную длину и подстроку
                if (right - left + 1 < minLen) {
                    minLen = right - left + 1;
                    minStr = s.substring(left, right + 1);
                }

                char leftChar = s.charAt(left);
                int leftCharCount = window.get(leftChar); // Сохраняем значение для избежания повторных вызовов getOrDefault
                window.put(leftChar, leftCharCount - 1);

                // Если символ leftChar есть в t и его частота в окне стала меньше требуемой
                if (need.containsKey(leftChar) && leftCharCount <= need.get(leftChar)) {
                    formed--;
                }
                left++; // Сдвигаем левую границу окна
            }
            right++; // Сдвигаем правую границу окна
        }
        return minStr;
    }

    public static void main(String[] args) {
        String s = "ADOBECODEBANC";
        String t = "ABC";
        Solution sol = new Solution();
        String result = sol.minWindow(s, t);
        System.out.println(result); 
    }
}

File: task1/.kt
fun minWindowSubstring(s: String, t: String): String { // Ищет минимальное окно в строке s, содержащее все символы строки t
    if (t.isEmpty()) return "" // Если t пустая строка, возвращает пустую строку

    val need = mutableMapOf<Char, Int>() // символ -> необходимое количество в t
    val window = mutableMapOf<Char, Int>() // символ -> количество в текущем окне

    for (char in t) { // Заполняем карту need
        need[char] = need.getOrDefault(char, 0) + 1 // Увеличиваем количество нужного символа на 1
    }

    var needCount = need.size // Количество уникальных символов в t
    var formed = 0 // Количество уникальных символов в t, содержащихся в окне в нужном количестве

    var left = 0 // Левая граница 
    var right = 0 // Правая граница 

    var minLen = Int.MAX_VALUE // Минимальная длина найденного окна (инициализируем максимальным значением)
    var minStr = "" // Минимальное окно (изначально пустое)

    while (right < s.length) { // Двигаем правую границу окна по строке s
        val char = s[right] // Текущий символ справа
        window[char] = window.getOrDefault(char, 0) + 1 // Увеличиваем количество символа в окне на 1

        if (need.containsKey(char) && window[char] == need[char]) { // Если символ нужен, и его количество в окне соответствует нужному
            formed++ // Увеличиваем счетчик сформированных символов
        }

        while (formed == needCount) { // Пока все нужные символы есть в окне в нужном количестве
            if (right - left + 1 < minLen) { // Если текущая длина окна меньше минимальной найденной
                minLen = right - left + 1 // Обновляем минимальную длину
                minStr = s.substring(left, right + 1) // Обновляем минимальное окно
            }

            val leftChar = s[left] // Символ слева
            window[leftChar] = window.getOrDefault(leftChar, 0) - 1 // Уменьшаем количество символа в окне на 1

            if (need.containsKey(leftChar) && window[leftChar]!! < need[leftChar]!!) { // Если символ нужен, и его количество стало меньше нужного
                formed-- // Уменьшаем счетчик сформированных символов
            }
            left++ // Сдвигаем левую границу окна
        }
        right++ // Сдвигаем правую границу окна
    }
    return minStr // Возвращает минимальное окно
}

fun main() { // Пример использования
    val s = "ADOBECODEBANC"
    val t = "ABC"
    val result = minWindowSubstring(s, t)
    println(result) 
}
